File: 1300-sum-of-mutated-array-closest-to-target/1300-sum-of-mutated-array-closest-to-target.c
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
#define MIN(a, b) (((a) < (b)) ? (a) : (b))

int findBestValue(int* arr, int arrSize, int target){
    int max = INT_MIN, total = 0;
    
    for(int i = 0; i < arrSize; i++){
        max = MAX(max, arr[i]);
        total += arr[i];
    }
        
    
    if(total <= target)
        return max;
    
    
    int left = 0, right = max, ans = INT_MAX, diff = INT_MAX;
    while(left <= right){
        int mid = (left + right) >> 1;
        int localSum = 0;
        for(int i = 0; i < arrSize; i++)
            localSum += (arr[i] > mid) ? mid : arr[i];
        
        if(abs(localSum - target) < diff){
            ans = mid;
            diff = abs(localSum - target);
        }else if(abs(localSum - target) == diff)
            ans = MIN(ans, mid);
        if(localSum > target){
            right = mid - 1;
        }
        else{
            left = mid + 1;
        }
    }
    return ans;
}

File: 1300-sum-of-mutated-array-closest-to-target/1300-sum-of-mutated-array-closest-to-target.py
class Solution:
    def findBestValue(self, arr: List[int], target: int) -> int:
        arr.sort(reverse = True)
        maxElem = arr[0]
        n = len(arr)
        while arr and arr[-1] * len(arr) < target:
            target -= arr.pop()
        
        if not arr:
            return maxElem
        if target / len(arr) - target // len(arr) <= 0.5:
            return target // len(arr)
        return target // len(arr) + 1